File: backend/apps/app_builder/app_builder.py
"""OpenSwarmApp Builder SubApp — CRUD, app management, file serving, and execution."""

import json
import mimetypes
import os
import shutil
from datetime import datetime
from contextlib import asynccontextmanager
from typing import Optional, Any
from pydantic import BaseModel

from fastapi import HTTPException
from fastapi.responses import Response

from backend.config.Apps import SubApp
from backend.core.db.PydanticStore import PydanticStore
from backend.apps.app_builder.OpenSwarmApp import OpenSwarmApp
from backend.config.paths import DB_ROOT
from backend.apps.app_builder.templates.templates import APP_BUILDER_SKILL, APP_BUILDER_TEMPLATE_FILES
from backend.apps.app_builder.utils.execute_backend_code import execute_backend_code, BackendExecResult
from backend.apps.app_builder.utils.walk_directory import walk_directory

APP_BUILDER_METADATA_DIR = os.path.join(DB_ROOT, "app_builder_metadata")
APP_BUILDER_CONTENT_DIR = os.path.join(APP_BUILDER_METADATA_DIR, "app_builder_content")

@asynccontextmanager
async def app_builder_lifespan():
    os.makedirs(APP_BUILDER_METADATA_DIR, exist_ok=True)
    os.makedirs(APP_BUILDER_CONTENT_DIR, exist_ok=True)
    yield


app_builder = SubApp("app_builder", app_builder_lifespan)

_store = PydanticStore[OpenSwarmApp](model_cls=OpenSwarmApp, data_dir=APP_BUILDER_METADATA_DIR, not_found_detail="OpenSwarmApp not found")


# ---------------------------------------------------------------------------
# File serving
# ---------------------------------------------------------------------------

@app_builder.router.get("/app/{app_id}/source_dir/{filepath:path}")
async def serve_app_file(app_id: str, filepath: str):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    full_path = os.path.normpath(os.path.join(folder, filepath))
    if not full_path.startswith(os.path.normpath(folder)):
        raise HTTPException(status_code=403, detail="Path traversal not allowed")
    if not os.path.isfile(full_path):
        raise HTTPException(status_code=404, detail="File not found")
    with open(full_path) as f:
        content = f.read()
    mime, _ = mimetypes.guess_type(filepath)
    return Response(content=content, media_type=mime or "text/plain")


@app_builder.router.get("/{app_id}/serve/{filepath:path}")
async def serve_app_file(app_id: str, filepath: str):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    full_path = os.path.normpath(os.path.join(folder, filepath))
    if not full_path.startswith(os.path.normpath(folder)):
        raise HTTPException(status_code=403, detail="Path traversal not allowed")
    if not os.path.isfile(full_path):
        raise HTTPException(status_code=404, detail="File not found in app")
    with open(full_path) as f:
        content = f.read()
    mime, _ = mimetypes.guess_type(filepath)
    return Response(content=content, media_type=mime or "text/plain")


# ---------------------------------------------------------------------------
# Workspace management
# ---------------------------------------------------------------------------

@app_builder.router.get("/app/{app_id}")
async def read_app(app_id: str):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    if not os.path.isdir(folder):
        raise HTTPException(status_code=404, detail="Workspace not found")
    files = walk_directory(folder)
    meta = None
    if "meta.json" in files:
        try:
            meta = json.loads(files["meta.json"])
        except (json.JSONDecodeError, ValueError):
            pass
    return {"files": files, "meta": meta}


class OpenSwarmAppSeedRequest(BaseModel):
    app_id: str
    files: Optional[dict[str, str]] = None
    meta: Optional[dict[str, Any]] = None

@app_builder.router.post("/app/seed")
async def seed_app(body: OpenSwarmAppSeedRequest):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, body.app_id)
    os.makedirs(folder, exist_ok=True)
    if body.files:
        for rel_path, content in body.files.items():
            full_path = os.path.normpath(os.path.join(folder, rel_path))
            if not full_path.startswith(os.path.normpath(folder)):
                continue
            os.makedirs(os.path.dirname(full_path), exist_ok=True)
            with open(full_path, "w") as f:
                f.write(content)
    else:
        for rel_path, content in APP_BUILDER_TEMPLATE_FILES.items():
            full_path = os.path.join(folder, rel_path)
            with open(full_path, "w") as f:
                f.write(content)
    with open(os.path.join(folder, "SKILL.md"), "w") as f:
        f.write(APP_BUILDER_SKILL)
    if body.meta:
        with open(os.path.join(folder, "meta.json"), "w") as f:
            json.dump(body.meta, f, indent=2)
    return {"path": os.path.abspath(folder)}


@app_builder.router.put("/app/{app_id}/file/{filepath:path}")
async def write_app_file(app_id: str, filepath: str, body: dict):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    if not os.path.isdir(folder):
        raise HTTPException(status_code=404, detail="Workspace not found")
    full_path = os.path.normpath(os.path.join(folder, filepath))
    if not full_path.startswith(os.path.normpath(folder)):
        raise HTTPException(status_code=403, detail="Path traversal not allowed")
    os.makedirs(os.path.dirname(full_path), exist_ok=True)
    with open(full_path, "w") as f:
        f.write(body.get("content", ""))
    return {"ok": True}


@app_builder.router.delete("/app/{app_id}/file/{filepath:path}")
async def delete_app_file(app_id: str, filepath: str):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    if not os.path.isdir(folder):
        raise HTTPException(status_code=404, detail="Workspace not found")
    full_path = os.path.normpath(os.path.join(folder, filepath))
    if not full_path.startswith(os.path.normpath(folder)):
        raise HTTPException(status_code=403, detail="Path traversal not allowed")
    if os.path.isfile(full_path):
        os.remove(full_path)
        parent = os.path.dirname(full_path)
        while parent != os.path.normpath(folder):
            if os.path.isdir(parent) and not os.listdir(parent):
                os.rmdir(parent)
                parent = os.path.dirname(parent)
            else:
                break
    return {"ok": True}


# ---------------------------------------------------------------------------
# OpenSwarmApp CRUD
# ---------------------------------------------------------------------------

@app_builder.router.get("/list")
async def list_apps():
    return {"apps": [o.model_dump() for o in _store.load_all()]}


@app_builder.router.get("/{app_id}")
async def get_app(app_id: str):
    return _store.load(app_id).model_dump()


class OpenSwarmAppCreate(BaseModel):
    name: str
    description: str = ""
    icon: str = "view_quilt"
    files: Optional[dict[str, str]] = None
    thumbnail: Optional[str] = None

@app_builder.router.post("/create")
async def create_app(body: OpenSwarmAppCreate):
    now = datetime.now().isoformat()
    app = OpenSwarmApp(
        name=body.name, description=body.description, icon=body.icon,
        thumbnail=body.thumbnail, created_at=now, updated_at=now,
    )
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app.id)
    os.makedirs(folder, exist_ok=True)
    seed_files = body.files if body.files else APP_BUILDER_TEMPLATE_FILES
    for rel_path, content in seed_files.items():
        full_path = os.path.normpath(os.path.join(folder, rel_path))
        if not full_path.startswith(os.path.normpath(folder)):
            continue
        os.makedirs(os.path.dirname(full_path), exist_ok=True)
        with open(full_path, "w") as f:
            f.write(content)
    with open(os.path.join(folder, "SKILL.md"), "w") as f:
        f.write(APP_BUILDER_SKILL)
    _store.save(app)
    return {"ok": True, "app": app.model_dump()}


class OpenSwarmAppUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    icon: Optional[str] = None
    thumbnail: Optional[str] = None

@app_builder.router.put("/{app_id}")
async def update_app(app_id: str, body: OpenSwarmAppUpdate):
    app = _store.load(app_id)
    for k, v in body.model_dump(exclude_none=True).items():
        setattr(app, k, v)
    app.updated_at = datetime.now().isoformat()
    _store.save(app)
    return {"ok": True, "app": app.model_dump()}


@app_builder.router.delete("/{app_id}")
async def delete_app(app_id: str):
    folder = os.path.join(APP_BUILDER_CONTENT_DIR, app_id)
    if os.path.isdir(folder):
        shutil.rmtree(folder)
    _store.delete(app_id)
    return {"ok": True}


# ---------------------------------------------------------------------------
# Execution
# ---------------------------------------------------------------------------

class OpenSwarmAppExecute(BaseModel):
    app_id: str

class OpenSwarmAppExecuteResult(BaseModel):
    app_id: str
    app_name: str
    frontend_code: str
    backend_result: Optional[dict[str, Any]] = None
    stdout: Optional[str] = None
    stderr: Optional[str] = None
    error: Optional[str] = None

def _read_app_file(app_id: str, filename: str) -> Optional[str]:
    path = os.path.join(APP_BUILDER_CONTENT_DIR, app_id, filename)
    if not os.path.isfile(path):
        return None
    with open(path) as f:
        return f.read()

@app_builder.router.post("/execute")
async def execute_app(body: OpenSwarmAppExecute):
    backend_code = _read_app_file(body.app_id, "backend.py")
    backend_result = None
    stdout_text = None
    stderr_text = None
    error = None
    app_name: Optional[str] = json.loads(_read_app_file(body.app_id, "meta.json")).get("name", "")
    assert app_name is not None, "App name is required but not found in meta.json"

    if backend_code:
        try:
            exec_result: BackendExecResult = await execute_backend_code(backend_code)
            backend_result = exec_result.result
            stdout_text = exec_result.stdout
            stderr_text = exec_result.stderr
        except Exception as e:
            error = str(e)

    return OpenSwarmAppExecuteResult(
        app_id=body.app_id, app_name=app_name,
        frontend_code=_read_app_file(body.app_id, "index.html") or "",
        backend_result=backend_result, stdout=stdout_text,
        stderr=stderr_text, error=error,
    ).model_dump()

File: frontend/src/shared/backend-bridge/apps/app_builder.ts
import { createAsyncThunk } from '@reduxjs/toolkit';
import { API_BASE } from '@/shared/backend-bridge/base_routes';

const APP_BUILDER_API: string = `${API_BASE}/app_builder`;

// ---------------------------------------------------------------------------
// File serving
// ---------------------------------------------------------------------------


const serve_app_source_file_endpoint: string = `${APP_BUILDER_API}/app/source_dir`;
async function serve_app_source_file_function(payload: {
  appId: string;
  filepath: string;
}): Promise<string> {
  const res = await fetch(`${APP_BUILDER_API}/app/${payload.appId}/source_dir/${payload.filepath}`, {
    method: 'GET',
  });
  const text = await res.text();
  return text;
}
export const SERVE_APP_SOURCE_FILE = createAsyncThunk(
  serve_app_source_file_endpoint,
  serve_app_source_file_function,
);


const serve_app_file_endpoint: string = `${APP_BUILDER_API}/serve`;
async function serve_app_file_function(payload: {
  appId: string;
  filepath: string;
}): Promise<string> {
  const res = await fetch(`${APP_BUILDER_API}/${payload.appId}/serve/${payload.filepath}`, {
    method: 'GET',
  });
  const text = await res.text();
  return text;
}
export const SERVE_APP_FILE = createAsyncThunk(
  serve_app_file_endpoint,
  serve_app_file_function,
);


// ---------------------------------------------------------------------------
// Workspace management
// ---------------------------------------------------------------------------


const read_app_endpoint: string = `${APP_BUILDER_API}/app`;
async function read_app_function(appId: string): Promise<{ files: Record<string, string>; meta: Record<string, unknown> | null }> {
  const res = await fetch(`${APP_BUILDER_API}/app/${appId}`, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as { files: Record<string, string>; meta: Record<string, unknown> | null };
}
export const READ_APP = createAsyncThunk(
  read_app_endpoint,
  read_app_function,
);


const seed_app_endpoint: string = `${APP_BUILDER_API}/app/seed`;
async function seed_app_function(body: {
  app_id: string;
  files?: Record<string, string> | null;
  meta?: Record<string, unknown> | null;
}): Promise<{ path: string }> {
  const res = await fetch(seed_app_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const data = await res.json();
  return data as { path: string };
}
export const SEED_APP = createAsyncThunk(
  seed_app_endpoint,
  seed_app_function,
);


const write_app_file_endpoint: string = `${APP_BUILDER_API}/app/file`;
async function write_app_file_function(payload: {
  appId: string;
  filepath: string;
  content: string;
}): Promise<{ ok: boolean }> {
  const res = await fetch(`${APP_BUILDER_API}/app/${payload.appId}/file/${payload.filepath}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ content: payload.content }),
  });
  const data = await res.json();
  return data as { ok: boolean };
}
export const WRITE_APP_FILE = createAsyncThunk(
  write_app_file_endpoint,
  write_app_file_function,
);


const delete_app_file_endpoint: string = `${APP_BUILDER_API}/app/file/delete`;
async function delete_app_file_function(payload: {
  appId: string;
  filepath: string;
}): Promise<{ ok: boolean }> {
  const res = await fetch(`${APP_BUILDER_API}/app/${payload.appId}/file/${payload.filepath}`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as { ok: boolean };
}
export const DELETE_APP_FILE = createAsyncThunk(
  delete_app_file_endpoint,
  delete_app_file_function,
);


// ---------------------------------------------------------------------------
// App CRUD
// ---------------------------------------------------------------------------


const list_apps_endpoint: string = `${APP_BUILDER_API}/list`;
async function list_apps_function(): Promise<{ apps: Record<string, unknown>[] }> {
  const res = await fetch(list_apps_endpoint, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as { apps: Record<string, unknown>[] };
}
export const LIST_APPS = createAsyncThunk(
  list_apps_endpoint,
  list_apps_function,
);


const get_app_endpoint: string = `${APP_BUILDER_API}/get`;
async function get_app_function(appId: string): Promise<Record<string, unknown>> {
  const res = await fetch(`${APP_BUILDER_API}/${appId}`, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as Record<string, unknown>;
}
export const GET_APP = createAsyncThunk(
  get_app_endpoint,
  get_app_function,
);


const create_app_endpoint: string = `${APP_BUILDER_API}/create`;
async function create_app_function(body: {
  name: string;
  description?: string;
  icon?: string;
  files?: Record<string, string> | null;
  thumbnail?: string | null;
}): Promise<{ ok: boolean; app: Record<string, unknown> }> {
  const res = await fetch(create_app_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const data = await res.json();
  return data as { ok: boolean; app: Record<string, unknown> };
}
export const CREATE_APP = createAsyncThunk(
  create_app_endpoint,
  create_app_function,
);


const update_app_endpoint: string = `${APP_BUILDER_API}/update`;
async function update_app_function(payload: {
  appId: string;
  name?: string;
  description?: string;
  icon?: string;
  thumbnail?: string | null;
}): Promise<{ ok: boolean; app: Record<string, unknown> }> {
  const { appId, ...updates } = payload;
  const res = await fetch(`${APP_BUILDER_API}/${appId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updates),
  });
  const data = await res.json();
  return data as { ok: boolean; app: Record<string, unknown> };
}
export const UPDATE_APP = createAsyncThunk(
  update_app_endpoint,
  update_app_function,
);


const delete_app_endpoint: string = `${APP_BUILDER_API}/delete`;
async function delete_app_function(appId: string): Promise<{ ok: boolean }> {
  const res = await fetch(`${APP_BUILDER_API}/${appId}`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as { ok: boolean };
}
export const DELETE_APP = createAsyncThunk(
  delete_app_endpoint,
  delete_app_function,
);


// ---------------------------------------------------------------------------
// Execution
// ---------------------------------------------------------------------------


const execute_app_endpoint: string = `${APP_BUILDER_API}/execute`;
async function execute_app_function(appId: string): Promise<{
  app_id: string;
  app_name: string;
  frontend_code: string;
  backend_result: Record<string, unknown> | null;
  stdout: string | null;
  stderr: string | null;
  error: string | null;
}> {
  const res = await fetch(execute_app_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ app_id: appId }),
  });
  const data = await res.json();
  return data as {
    app_id: string;
    app_name: string;
    frontend_code: string;
    backend_result: Record<string, unknown> | null;
    stdout: string | null;
    stderr: string | null;
    error: string | null;
  };
}
export const EXECUTE_APP = createAsyncThunk(
  execute_app_endpoint,
  execute_app_function,
);
